ramanujan/matrix.py

```diff
@@ -48,21 +48,42 @@ def walk(self, trajectory, iterations, start):
         where `M=self`, `(t_0, ..., t_k)=trajectory`, `n=iterations` and `(s_0, ..., s_k)=start`.
 
         This is a generalization of the basic (and most common) case $\prod_{i=0}^{n-1}M(s+i)$,
-        where `M=self`, `n=iterations` and `s=start`.
+        where M=self, n=iterations and s=start.
 
         Args:
             trajectory: the trajectory of a single step in the walk, as defined above.
-            iterations: the amount of multiplications to perform
+            iterations: The amount of multiplications to perform or a list of integers representing indexes along the multipication proccess to be returned. 
+                        If it's a list, the iterations will be sorted in ascending order to improve efficiency. 
             start: the starting point of the matrix multiplication
         Returns:
-            the walk multiplication as defined above.
+            Steps from the walk multiplication as defined above.
         """
+
         position = start
-        retval = Matrix.eye(2)
-        for _ in range(iterations):
-            retval *= self.subs(position)
+        value_matrix = Matrix.eye(2)
+
+        retval = [value_matrix]
+
+
+        islist = isinstance(iterations, list)
+
+        if islist: iterations.sort()
```

Comment:
You generally want to deal with all the type issues at the beginning.
Some approaches that are simpler and less bug-prone:
1. `iterations = [iterations] if type(iterations) is not List else iterations`
2. Use an inner walk function that receives only 1 value of iterations, and call it `iterations` time, each time starting from the previous point

Reply:
Also, what do you think about returning a single value if iterations is not a list? That way we could mimic a similar behavior of function overloads in c/c++. sympy works like this in a lot of places.